### PR TITLE
glib: Do not use ptr::offset/offset_from for private/impl offset

### PR DESCRIPTION
**Pull request by heftig**
Doing this violates Rust's [strict provenance rules][1]. Unfortunately, the API for doing this properly ([`ptr::expose_addr`][2]) isn't ready yet. However, we can work around this on stable Rust by casting through `isize` for offset calculations.

[1]: https://doc.rust-lang.org/std/ptr/index.html
[2]: https://doc.rust-lang.org/std/primitive.pointer.html#method.expose_addr

### CODE REVIEW

**Comment by heftig** [comment]
Should fix [gstreamer-rs#470](https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/-/issues/470).

File: glib/src/subclass/types.rs [hunk 1]
```diff
@@ -95,7 +95,7 @@ unsafe impl<T: InstanceStruct> InstanceStructExt for T {
             let data = Self::Type::type_data();
             let private_offset = data.as_ref().impl_offset();
             let ptr: *const u8 = self as *const _ as *const u8;
-            let imp_ptr = ptr.offset(private_offset);
+            let imp_ptr = ((ptr as isize) + private_offset) as *const u8;
             let imp = imp_ptr as *const Self::Type;
```
Comment (sdroege) [inline glib/src/subclass/types.rs:99]: I think this is still wrong. There should be no `*const u8` anywhere here, and wherever it is used you should use `isize` or `usize`.

**Comment by sdroege** [comment]
Should also make a 0.16 bugfix release with this.

**Comment by sdroege** [comment]
Also causes https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs/-/issues/373 . Quite a horrible bug :)

File: glib/src/subclass/types.rs [hunk 2]
```diff
@@ -95,7 +95,7 @@ unsafe impl<T: InstanceStruct> InstanceStructExt for T {
             let data = Self::Type::type_data();
             let private_offset = data.as_ref().impl_offset();
             let ptr: *const u8 = self as *const _ as *const u8;
-            let imp_ptr = ptr.offset(private_offset);
+            let imp_ptr = ((ptr as isize) + private_offset) as *const u8;
```
Comment (GuillaumeGomez) [inline glib/src/subclass/types.rs:98]: Could you add comments explaining why we do it like and also a note saying that once `expose_addr` is stable, we should use it instead please?

Comment (sdroege) [reply glib/src/subclass/types.rs:98]: Already done locally

File: glib/src/subclass/types.rs [hunk 3]
```diff
@@ -88,16 +88,57 @@ pub unsafe trait InstanceStructExt: InstanceStruct {
     fn class(&self) -> &<Self::Type as ObjectSubclass>::Class;
 }
 
+// rustdoc-stripper-ignore-next
+/// Offset `ptr` by `offset` *bytes* and cast the result to `*const U`.
```
Comment (GuillaumeGomez) [inline glib/src/subclass/types.rs:92]: That's much better, thanks a lot!

**Pull request merged**